File: src/integration/jvm/src/jvmTest/java/community/flock/wirespec/integration/jvm/java/HttpClientTransportationTest.java
package community.flock.wirespec.integration.jvm.java;

import community.flock.wirespec.java.Wirespec;
import org.junit.jupiter.api.Test;
import org.mockito.ArgumentCaptor;
import org.mockito.MockedStatic;

import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.concurrent.CompletableFuture;

import static org.junit.jupiter.api.Assertions.assertArrayEquals;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.mockStatic;
import static org.mockito.Mockito.when;

class HttpClientTransportationTest {

    @Test
    void testTransport() throws Exception {
        HttpClient mockClient = mock(HttpClient.class);
        HttpResponse<byte[]> mockResponse = mock(HttpResponse.class);

        String body = "{\"hello\":\"world\"}";
        when(mockResponse.statusCode()).thenReturn(200);
        when(mockResponse.body()).thenReturn(body.getBytes());
        when(mockResponse.headers()).thenReturn(java.net.http.HttpHeaders.of(Collections.emptyMap(), (s1, s2) -> true));

        when(mockClient.sendAsync(any(HttpRequest.class), any(HttpResponse.BodyHandler.class)))
                .thenReturn(CompletableFuture.completedFuture(mockResponse));

        HttpClientTransportation transportation = new HttpClientTransportation("https://67.com", mockClient);
        Wirespec.RawRequest request = new Wirespec.RawRequest(
                "POST",
                List.of("path", "to", "resource"),
                Map.of("query", List.of("v1")),
                Map.of("Content-Type", List.of("application/json")),
                body.getBytes()
        );

        Wirespec.RawResponse response = transportation.transport(request).get();

        ArgumentCaptor<HttpRequest> requestCaptor = ArgumentCaptor.forClass(HttpRequest.class);
        org.mockito.Mockito.verify(mockClient).sendAsync(requestCaptor.capture(), any());

        HttpRequest capturedRequest = requestCaptor.getValue();
        assertEquals("https://67.com/path/to/resource?query=v1", capturedRequest.uri().toString());
        assertEquals("POST", capturedRequest.method());
        assertEquals("application/json", capturedRequest.headers().firstValue("Content-Type").orElse(null));

        assertEquals(200, response.statusCode());
        assertArrayEquals(body.getBytes(), response.body());
    }

    @Test
    void testTransportTrailingSlash() throws Exception {
        HttpClient mockClient = mock(HttpClient.class);
        HttpResponse<byte[]> mockResponse = mock(HttpResponse.class);

        String body = "{\"hello\":\"world\"}";
        when(mockResponse.statusCode()).thenReturn(200);
        when(mockResponse.body()).thenReturn(body.getBytes());
        when(mockResponse.headers()).thenReturn(java.net.http.HttpHeaders.of(Collections.emptyMap(), (s1, s2) -> true));

        when(mockClient.sendAsync(any(HttpRequest.class), any(HttpResponse.BodyHandler.class)))
                .thenReturn(CompletableFuture.completedFuture(mockResponse));

        HttpClientTransportation transportation = new HttpClientTransportation("https://67.com/", mockClient);
        Wirespec.RawRequest request = new Wirespec.RawRequest(
                "GET",
                List.of("path", "to", "resource"),
                Collections.emptyMap(),
                Collections.emptyMap(),
                null
        );

        transportation.transport(request).get();

        ArgumentCaptor<HttpRequest> requestCaptor = ArgumentCaptor.forClass(HttpRequest.class);
        org.mockito.Mockito.verify(mockClient).sendAsync(requestCaptor.capture(), any());

        HttpRequest capturedRequest = requestCaptor.getValue();
        assertEquals("https://67.com/path/to/resource", capturedRequest.uri().toString());
    }

    @Test
    void testTransportStaticMock() throws Exception {
        try (MockedStatic<HttpClient> httpClientMockedStatic = mockStatic(HttpClient.class)) {
            HttpClient.Builder mockBuilder = mock(HttpClient.Builder.class);
            HttpClient mockClient = mock(HttpClient.class);
            HttpResponse<byte[]> mockResponse = mock(HttpResponse.class);

            httpClientMockedStatic.when(HttpClient::newBuilder).thenReturn(mockBuilder);
            when(mockBuilder.build()).thenReturn(mockClient);

            String body = "{\"hello\":\"world\"}";
            when(mockResponse.statusCode()).thenReturn(200);
            when(mockResponse.body()).thenReturn(body.getBytes());
            when(mockResponse.headers()).thenReturn(java.net.http.HttpHeaders.of(Collections.emptyMap(), (s1, s2) -> true));

            when(mockClient.sendAsync(any(HttpRequest.class), any(HttpResponse.BodyHandler.class)))
                    .thenReturn(CompletableFuture.completedFuture(mockResponse));

            HttpClientTransportation transportation = new HttpClientTransportation("https://67.com");
            Wirespec.RawRequest request = new Wirespec.RawRequest(
                    "POST",
                    List.of("path", "to", "resource"),
                    Map.of("query", List.of("v1")),
                    Map.of("Content-Type", List.of("application/json")),
                    body.getBytes()
            );

            Wirespec.RawResponse response = transportation.transport(request).get();

            assertEquals(200, response.statusCode());
            assertArrayEquals(body.getBytes(), response.body());
        }
    }
}

File: src/integration/jvm/src/jvmTest/kotlin/community/flock/wirespec/integration/jvm/kotlin/HttpClientTransportationTest.kt
package community.flock.wirespec.integration.jvm.kotlin

import community.flock.wirespec.kotlin.Wirespec
import io.mockk.coEvery
import io.mockk.mockk
import io.mockk.slot
import io.mockk.verify
import kotlinx.coroutines.runBlocking
import org.junit.jupiter.api.Assertions.assertArrayEquals
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test
import java.net.http.HttpClient
import java.net.http.HttpHeaders
import java.net.http.HttpRequest
import java.net.http.HttpResponse
import java.util.concurrent.CompletableFuture

class HttpClientTransportationTest {

    @Test
    fun testTransport() {
        runBlocking {
            val mockClient = mockk<HttpClient>()
            val mockResponse = mockk<HttpResponse<ByteArray>>()

            val body = "{\"hello\":\"world\"}"
            coEvery { mockResponse.statusCode() } returns 200
            coEvery { mockResponse.body() } returns body.toByteArray()
            coEvery { mockResponse.headers() } returns HttpHeaders.of(emptyMap()) { _, _ -> true }

            coEvery { mockClient.sendAsync(any<HttpRequest>(), any<HttpResponse.BodyHandler<ByteArray>>()) } returns CompletableFuture.completedFuture(mockResponse)

            val transportation = HttpClientTransportation("https://67.com", mockClient)
            val request = Wirespec.RawRequest(
                method = "POST",
                path = listOf("path", "to", "resource"),
                queries = mapOf("query" to listOf("v1")),
                headers = mapOf("Content-Type" to listOf("application/json")),
                body = body.toByteArray(),
            )

            val response = transportation.transport(request)

            val requestSlot = slot<HttpRequest>()
            verify { mockClient.sendAsync(capture(requestSlot), any<HttpResponse.BodyHandler<ByteArray>>()) }

            val capturedRequest = requestSlot.captured
            assertEquals("https://67.com/path/to/resource?query=v1", capturedRequest.uri().toString())
            assertEquals("POST", capturedRequest.method())
            assertEquals("application/json", capturedRequest.headers().firstValue("Content-Type").orElse(null))

            assertEquals(200, response.statusCode)
            assertArrayEquals(body.toByteArray(), response.body)
        }
    }

    @Test
    fun testTransportWithTrailingSlashAndEmptySegments() {
        runBlocking {
            val mockClient = mockk<HttpClient>()
            val mockResponse = mockk<HttpResponse<ByteArray>>()

            coEvery { mockResponse.statusCode() } returns 200
            coEvery { mockResponse.body() } returns ByteArray(0)
            coEvery { mockResponse.headers() } returns HttpHeaders.of(emptyMap()) { _, _ -> true }

            coEvery { mockClient.sendAsync(any<HttpRequest>(), any<HttpResponse.BodyHandler<ByteArray>>()) } returns CompletableFuture.completedFuture(mockResponse)

            val transportation = HttpClientTransportation("https://67.com/", mockClient)
            val request = Wirespec.RawRequest(
                method = "GET",
                path = listOf("", "path", "", "resource", ""),
                queries = emptyMap(),
                headers = emptyMap(),
                body = null,
            )

            transportation.transport(request)

            val requestSlot = slot<HttpRequest>()
            verify { mockClient.sendAsync(capture(requestSlot), any<HttpResponse.BodyHandler<ByteArray>>()) }

            val capturedRequest = requestSlot.captured
            assertEquals("https://67.com/path/resource", capturedRequest.uri().toString())
        }
    }
}
